server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -134,6 +162,238 @@ public List<MetricAggregatorInfo> generateMetricAggregatorInfos(MapperService ma
         return metricAggregatorInfos;
     }
 
+    /**
+     * Generates the configuration required to perform aggregation for all the metrics on a field
+     *
+     * @return list of MetricAggregatorInfo
+     */
+    public List<SequentialDocValuesIterator> getMetricReaders(SegmentWriteState state, Map<String, DocValuesProducer> fieldProducerMap)
+        throws IOException {
+
+        List<SequentialDocValuesIterator> metricReaders = new ArrayList<>();
+        for (Metric metric : this.starTreeField.getMetrics()) {
+            for (MetricStat metricStat : metric.getMetrics()) {
+                SequentialDocValuesIterator metricReader;
+                FieldInfo metricFieldInfo = state.fieldInfos.fieldInfo(metric.getField());
+                if (metricStat != MetricStat.COUNT) {
+                    metricReader = new SequentialDocValuesIterator(
+                        fieldProducerMap.get(metricFieldInfo.name).getSortedNumeric(metricFieldInfo)
+                    );
+                } else {
+                    metricReader = new SequentialDocValuesIterator();
```

Comment:
Major :
How does this work for merge , since for merge we need to read the actual values present in count dvs.

Apart from this, we need to replace all individual count fields with one count field - this one we can take separately as well.

Reply:
> How does this work for merge , since for merge we need to read the actual values present in count dvs.

Good point, I guess we then need to load the metrics behind the field.

> Apart from this, we need to replace all individual count fields with one count field - this one we can take separately as well.

Yeah, let's create an issue for this.

Reply:
> Good point, I guess we then need to load the metrics behind the field.

Can we do it only for merge ? for flush current code might work.

Reply:
> Apart from this, we need to replace all individual count fields with one count field - this one we can take separately as well.

This cannot be done looks like. Count for each metric depends on whether the metric field is present in document. So each metric will have different count.

Reply:
As discussed offline, the metric readers are fetched differently during flush and merge. While we initialize the emptySortedNumeric during flush, we use the [correct sorted numeric from doc values producer](https://github.com/opensearch-project/OpenSearch/pull/14809/files#diff-db4fa7700a22f6e2c0c786c24f066c223c9c892d6978644e436b91f73bbbe44dR346) during merges.